linkerd/proxy/transport/src/detect.rs

```diff
@@ -0,0 +1,163 @@
+    capacity: usize,
+    timeout: time::Duration,
+}
+
+#[derive(Copy, Clone)]
+pub struct DetectService<N, D, T> {
+    target: T,
+    new_accept: N,
+    detect: D,
+    capacity: usize,
+    timeout: time::Duration,
+}
+
+#[derive(Debug)]
+pub struct DetectTimeout {
+    bytes: usize,
+    elapsed: time::Duration,
+}
+
+// === impl NewDetectService ===
+
+impl<N, D: Clone> NewDetectService<N, D> {
+    const BUFFER_CAPACITY: usize = 8192;
+
+    pub fn new(timeout: time::Duration, new_accept: N, detect: D) -> Self {
+        Self {
+            detect,
+            new_accept,
+            capacity: Self::BUFFER_CAPACITY,
+            timeout,
+        }
+    }
+
+    pub fn layer(
+        timeout: time::Duration,
+        detect: D,
+    ) -> impl layer::Layer<N, Service = Self> + Clone {
+        layer::mk(move |new| Self::new(timeout, new, detect.clone()))
+    }
+}
+
+impl<N: Clone, D: Clone, T> NewService<T> for NewDetectService<N, D> {
+    type Service = DetectService<N, D, T>;
+
+    fn new_service(&mut self, target: T) -> DetectService<N, D, T> {
+        DetectService {
+            target,
+            new_accept: self.new_accept.clone(),
+            detect: self.detect.clone(),
+            capacity: self.capacity,
+            timeout: self.timeout,
+        }
+    }
+}
+
+// === impl DetectService ===
+
+impl<N, S, D, T, I> tower::Service<I> for DetectService<N, D, T>
+where
+    T: Clone + Send + 'static,
+    I: io::AsyncRead + Send + Unpin + 'static,
+    D: Detect,
+    D::Protocol: std::fmt::Debug,
+    N: NewService<(Option<D::Protocol>, T), Service = S> + Clone + Send + 'static,
+    S: tower::Service<io::PrefixedIo<I>, Response = ()> + Send,
+    S::Error: Into<Error>,
+    S::Future: Send,
+{
+    type Response = ();
+    type Error = Error;
+    type Future = Pin<Box<dyn Future<Output = Result<(), Error>> + Send + 'static>>;
+
+    fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(().into()))
+    }
+
+    fn call(&mut self, mut io: I) -> Self::Future {
+        let mut new_accept = self.new_accept.clone();
+        let mut buf = BytesMut::with_capacity(self.capacity);
+        let detect = self.detect.clone();
+        let target = self.target.clone();
+        let timeout = self.timeout;
+        Box::pin(async move {
+            trace!("Starting protocol detection");
+            let t0 = time::Instant::now();
+            let protocol = futures::select_biased! {
+                res = detect.detect(&mut io, &mut buf).fuse() => res?,
+                _ = time::sleep(timeout).fuse() => {
+                    let bytes = buf.len();
+                    let elapsed = time::Instant::now() - t0;
+                    return Err(DetectTimeout { bytes, elapsed }.into());
+                }
+            };
+            debug!(
+                ?protocol,
+                elapsed = ?(time::Instant::now() - t0),
+                "Detected"
+            );
```

Comment:
this is really lovely --- it's nice how async-await lets us express the timing code in a way that feels really natural. i feel like this would have been pretty unpleasant to do with manual future impls.